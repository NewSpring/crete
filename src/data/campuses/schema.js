import { Campus } from '@apollosproject/data-connector-rock';

export default Campus.schema;
