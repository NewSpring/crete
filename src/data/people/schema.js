import { Person } from '@apollosproject/data-connector-rock';
import gql from 'graphql-tag';

const peopleSchema = gql`
  ${Person.schema}

  extend type Person {
    impersonationParameter: String! @deprecated(reason: "No longer used.")
    isGroupLeader: Boolean
    isStaff: Boolean
  }

  extend type Query {
    forgotPasswordURL: String
  }
`;

export default peopleSchema;
