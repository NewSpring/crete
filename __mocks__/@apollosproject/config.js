const ApolloServer = require.requireActual('@apollosproject/config').default;

ApolloServer.loadJs({
  ROCK: {
    API_URL: 'https://apollosrock.newspring.cc/api',
    API_TOKEN: 'some-rock-token',
    IMAGE_URL: 'https://apollosrock.newspring.cc/GetImage.ashx',
    SHARE_URL: 'https://newspring.cc',
    TIMEZONE: 'America/New_York',
  },
  ROCK_CONSTANTS: {
    IMAGE: 10,
    AUDIO_FILE: 77,
    VIDEO_FILE: 79,
  },
  ROCK_MAPPINGS: {
    FEED_CONTENT_CHANNEL_IDS: [1, 2, 3, 4, 6, 8],
    SERIES_CONTENT_CHANNEL_TYPE_IDS: [6, 7],
    CONTENT_ITEM: {
      ContentSeriesContentItem: {
        EntityType: 'ContentChannelItem',
        ContentChannelId: [18, 20],
      },
      DevotionalContentItem: {
        EntityType: 'ContentChannelItem',
        ContentChannelId: [17],
      },
      MediaContentItem: {
        EntityType: 'ContentChannelItem',
      },
      UniversalContentItem: {
        EntityType: 'ContentChannelItem',
      },
      ContentItem: {
        EntityType: 'ContentChannelItem',
      },
      WeekendContentItem: {
        EntityType: 'ContentChannelItem',
      },
    },
  },
});

export default ApolloServer;
