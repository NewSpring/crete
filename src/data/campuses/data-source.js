import { Campus as apollosCampus } from '@apollosproject/data-connector-rock';
import bugsnagClient from '../../bugsnag';

export default class Campus extends apollosCampus.dataSource {
  getPublicByLocation = async (location) => {
    const allCampuses = await this.getByLocation(location);

    // check public attribute first
    const campuses = allCampuses.filter((campus) => {
      const { attributeValues: { public: { value } = {} } = {} } = campus;
      if (!value) {
        bugsnagClient.notify(
          new Error('cannot determine if campus is public'),
          {
            metaData: { campus },
            severity: 'warning',
          }
        );

        // fall back is filter out if there's no location or image
        return !!(
          campus.location.longitude &&
          campus.location.latitude &&
          campus.location.image
        );
      }
      return value === 'True';
    });
    return campuses;
  };
}
