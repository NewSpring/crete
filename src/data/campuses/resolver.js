import { Campus as originalCampus } from '@apollosproject/data-connector-rock';
import { resolverMerge } from '@apollosproject/server-core';
import { createAssetUrl } from '../utils';

const resolver = {
  Query: {
    campuses: (root, { location }, { dataSources }) =>
      dataSources.Campus.getPublicByLocation(location),
  },
  Campus: {
    image: ({ location }) =>
      location.image
        ? {
            uri: createAssetUrl(location.image),
          }
        : null,
  },
};

export default resolverMerge(resolver, originalCampus);
