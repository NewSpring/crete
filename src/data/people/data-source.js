import { Person as basePerson } from '@apollosproject/data-connector-rock';

export default class Person extends basePerson.dataSource {
  async isStaff(id) {
    const staff = await this.request('GroupMembers')
      // active and not archived
      .filter(
        "GroupId eq 3 and GroupMemberStatus eq '1' and IsArchived eq false"
      )
      .get();
    const staffIds = staff.map(({ personId }) => personId);
    if (!staffIds.includes(id)) return false;
    return true;
  }
}
