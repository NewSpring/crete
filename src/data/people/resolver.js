import { Person as originalPerson } from '@apollosproject/data-connector-rock';
import { resolverMerge } from '@apollosproject/server-core';
import ApollosConfig from '@apollosproject/config';

const { ROCK } = ApollosConfig;

const resolver = {
  Query: {
    forgotPasswordURL: () => ROCK.FORGOT_PASSWORD_URL,
  },
  Person: {
    // deprecated
    impersonationParameter: () => 'deprecatedtoken',
    isGroupLeader: async ({ id }, args, { dataSources: { Group } }) => {
      const groups = await Group.getByPerson({ personId: id, asLeader: true });
      return groups.length > 0;
    },
    isStaff: ({ id }, args, { dataSources: { Person } }) => Person.isStaff(id),
    photo: ({ photo: { path } }) => ({ uri: path }),
  },
};
export default resolverMerge(resolver, originalPerson);
