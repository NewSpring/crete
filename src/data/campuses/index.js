import { Campus } from '@apollosproject/data-connector-rock';
import resolver from './resolver';

const { schema, dataSource } = Campus;

export { schema, resolver, dataSource };
