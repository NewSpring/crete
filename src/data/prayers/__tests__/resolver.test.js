import { graphql } from 'graphql';
import { createTestHelpers } from '@apollosproject/server-core/lib/testUtils';
import { Followings, Auth } from '@apollosproject/data-connector-rock';
import {
  contentItemSchema,
  contentChannelSchema,
  themeSchema,
  scriptureSchema,
} from '@apollosproject/data-schema';

import * as Prayer from '../index';
import * as Person from '../../people';
import * as Campus from '../../campuses';

import oneRockPrayer, { twoRockPrayers } from '../../mocks/prayer';
import oneRockPerson from '../../mocks/person';
import oneRockCampus from '../../mocks/campus';

// define here any classes with dataSource functions you need to overwrite
const { getSchema, getContext } = createTestHelpers({
  Prayer,
  Auth,
  Person,
  Campus,
  Followings,
});

describe('Prayer resolver', () => {
  let schema;
  let context;
  let rootValue;
  beforeEach(() => {
    // define here any extra schemas necessary
    schema = getSchema([
      contentItemSchema,
      contentChannelSchema,
      themeSchema,
      scriptureSchema,
    ]);
    context = getContext();
    rootValue = {};
  });

  it('gets a prayer feed', async () => {
    const query = `
      query {
        prayerFeed {
          edges {
            node {
              ... on Prayer {
                id
                text
              }
            }
          }
        }
      }
    `;

    context.dataSources.Prayer.paginate = jest.fn(() =>
      Promise.resolve({
        edges: [{ node: { id: 1, text: 'very serious prayer' } }],
      })
    );
    context.dataSources.Auth.getCurrentPerson = jest.fn(() =>
      Promise.resolve(oneRockPerson)
    );
    const result = await graphql(schema, query, rootValue, context);
    expect(result).toMatchSnapshot();
  });

  it('gets all public prayer requests', async () => {
    const query = `
      query {
        prayers {
          id
          text
          campus {
            id
            name
          }
          startTime
          flagCount
          prayerCount
          isAnonymous
          isSaved
          requestor {
            id
            firstName
            lastName
          }
          person {
            id
          }
        }
      }
    `;

    context.dataSources.Prayer.getPrayers = jest.fn(() =>
      Promise.resolve(twoRockPrayers)
    );
    context.dataSources.Campus.getFromId = jest.fn(() =>
      Promise.resolve(oneRockCampus)
    );
    context.dataSources.Person.getFromAliasId = jest.fn(() =>
      Promise.resolve(oneRockPerson)
    );
    context.dataSources.Followings.getFollowingsForCurrentUserAndNode = jest.fn(
      () => Promise.resolve([{ text: 'fake prayer' }])
    );
    const result = await graphql(schema, query, rootValue, context);
    expect(result).toMatchSnapshot();
  });

  it('gets all public prayer requests by campus', async () => {
    const query = `
      query {
        campusPrayers {
          id
          text
          campus {
            id
            name
          }
        }
      }
    `;
    context.dataSources.Prayer.getPrayers = jest.fn(() =>
      Promise.resolve(twoRockPrayers)
    );
    context.dataSources.Campus.getFromId = jest.fn(() =>
      Promise.resolve(oneRockCampus)
    );
    const result = await graphql(schema, query, rootValue, context);
    expect(result).toMatchSnapshot();
  });

  it('gets all prayers from current person', async () => {
    const query = `
      query {
        userPrayers {
          id
          firstName
          text
        }
      }
    `;

    context.dataSources.Prayer.getPrayers = jest.fn(() =>
      Promise.resolve(twoRockPrayers)
    );
    const result = await graphql(schema, query, rootValue, context);
    expect(result).toMatchSnapshot();
  });

  it('gets all prayers from groups', async () => {
    const query = `
      query {
        groupPrayers {
          id
          text
        }
      }
    `;

    context.dataSources.Prayer.getPrayers = jest.fn(() =>
      Promise.resolve(twoRockPrayers)
    );
    const result = await graphql(schema, query, rootValue, context);
    expect(result).toMatchSnapshot();
  });
  it('gets all saved prayers', async () => {
    const query = `
      query {
        savedPrayers {
          id
          text
        }
      }
    `;

    context.dataSources.Prayer.getPrayers = jest.fn(() =>
      Promise.resolve(twoRockPrayers)
    );
    const result = await graphql(schema, query, rootValue, context);
    expect(result).toMatchSnapshot();
  });

  it('creates a new prayer', async () => {
    const query = `
      mutation {
        addPrayer(
          text: "Jesus Rocks"
          isAnonymous: true
        ) {
          id
          text
        }
      }
    `;
    context.dataSources.Prayer.add = jest.fn(() =>
      Promise.resolve(oneRockPrayer)
    );
    const result = await graphql(schema, query, rootValue, context);
    expect(result).toMatchSnapshot();
  });
  it('deletes a prayer', async () => {
    const query = `
      mutation {
        deletePrayer(
          nodeId: "Prayer:b36e55d803443431e96bb4b5068147ec"
        ) {
          id
          text
        }
      }
    `;
    context.dataSources.Prayer.deletePrayer = jest.fn(() =>
      Promise.resolve(oneRockPrayer)
    );
    const result = await graphql(schema, query, rootValue, context);
    expect(result).toMatchSnapshot();
  });

  it('increments prayed for a request', async () => {
    const query = `
      mutation {
        incrementPrayerCount(
          nodeId: "Prayer:b36e55d803443431e96bb4b5068147ec"
        ) {
          id
          text
        }
      }
    `;
    context.dataSources.Prayer.createInteraction = jest.fn(() => null);
    context.dataSources.Prayer.incrementPrayed = jest.fn(() =>
      Promise.resolve(oneRockPrayer)
    );
    const result = await graphql(schema, query, rootValue, context);
    expect(result).toMatchSnapshot();
  });

  it('flags a prayer request', async () => {
    const query = `
      mutation {
        flagPrayer(nodeId: "Prayer:b36e55d803443431e96bb4b5068147ec") {
          id
          text
        }
      }
    `;
    context.dataSources.Prayer.flag = jest.fn(() =>
      Promise.resolve(oneRockPrayer)
    );
    const result = await graphql(schema, query, rootValue, context);
    expect(result).toMatchSnapshot();
  });
  it('saves a prayer', async () => {
    const query = `
      mutation {
        savePrayer(nodeId: "Prayer:b36e55d803443431e96bb4b5068147ec") {
          id
          text
        }
      }
    `;
    context.dataSources.Followings.followNode = jest.fn(() => null);
    context.models.Node.get = jest.fn(() => Promise.resolve(oneRockPrayer));
    const result = await graphql(schema, query, rootValue, context);
    expect(result).toMatchSnapshot();
  });
  it('unsaves a prayer', async () => {
    const query = `
      mutation {
        unSavePrayer(nodeId: "Prayer:b36e55d803443431e96bb4b5068147ec") {
          id
          text
        }
      }
    `;
    context.dataSources.Followings.unFollowNode = jest.fn(() => null);
    context.models.Node.get = jest.fn(() => Promise.resolve(oneRockPrayer));
    const result = await graphql(schema, query, rootValue, context);
    expect(result).toMatchSnapshot();
  });
});
