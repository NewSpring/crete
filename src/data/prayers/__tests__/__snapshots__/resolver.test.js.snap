// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prayer resolver creates a new prayer 1`] = `
Object {
  "data": Object {
    "addPrayer": Object {
      "id": "Prayer:559b23fd0aa90e81b1c023e72e230fa1",
      "text": "Pray this works.",
    },
  },
}
`;

exports[`Prayer resolver deletes a prayer 1`] = `
Object {
  "data": Object {
    "deletePrayer": Object {
      "id": "Prayer:559b23fd0aa90e81b1c023e72e230fa1",
      "text": "Pray this works.",
    },
  },
}
`;

exports[`Prayer resolver flags a prayer request 1`] = `
Object {
  "data": Object {
    "flagPrayer": Object {
      "id": "Prayer:559b23fd0aa90e81b1c023e72e230fa1",
      "text": "Pray this works.",
    },
  },
}
`;

exports[`Prayer resolver gets a prayer feed 1`] = `
Object {
  "data": Object {
    "prayerFeed": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "Prayer:559b23fd0aa90e81b1c023e72e230fa1",
            "text": "very serious prayer",
          },
        },
      ],
    },
  },
}
`;

exports[`Prayer resolver gets all prayers from current person 1`] = `
Object {
  "data": Object {
    "userPrayers": Array [
      Object {
        "firstName": "Isaac",
        "id": "Prayer:559b23fd0aa90e81b1c023e72e230fa1",
        "text": "Pray this works.",
      },
      Object {
        "firstName": "Rich",
        "id": "Prayer:965b6e6d7046a885bea4e300b5c0400d",
        "text": "Help me",
      },
    ],
  },
}
`;

exports[`Prayer resolver gets all prayers from groups 1`] = `
Object {
  "data": Object {
    "groupPrayers": Array [
      Object {
        "id": "Prayer:559b23fd0aa90e81b1c023e72e230fa1",
        "text": "Pray this works.",
      },
      Object {
        "id": "Prayer:965b6e6d7046a885bea4e300b5c0400d",
        "text": "Help me",
      },
    ],
  },
}
`;

exports[`Prayer resolver gets all public prayer requests 1`] = `
Object {
  "data": Object {
    "prayers": Array [
      Object {
        "campus": Object {
          "id": "Campus:be35f49307d7297989d3514be788ef2d",
          "name": "Anderson",
        },
        "flagCount": 0,
        "id": "Prayer:559b23fd0aa90e81b1c023e72e230fa1",
        "isAnonymous": false,
        "isSaved": true,
        "person": Object {
          "id": "Person:057b0758543270d1cb21784cac4d3f5c",
        },
        "prayerCount": 4,
        "requestor": Object {
          "firstName": "Isaac",
          "id": "Person:057b0758543270d1cb21784cac4d3f5c",
          "lastName": "Hardy",
        },
        "startTime": "2019-07-02T13:08:02.035",
        "text": "Pray this works.",
      },
      Object {
        "campus": null,
        "flagCount": 1,
        "id": "Prayer:965b6e6d7046a885bea4e300b5c0400d",
        "isAnonymous": true,
        "isSaved": true,
        "person": Object {
          "id": "Person:057b0758543270d1cb21784cac4d3f5c",
        },
        "prayerCount": 0,
        "requestor": Object {
          "firstName": "Isaac",
          "id": "Person:057b0758543270d1cb21784cac4d3f5c",
          "lastName": "Hardy",
        },
        "startTime": "2019-07-03T13:08:02.035",
        "text": "Help me",
      },
    ],
  },
}
`;

exports[`Prayer resolver gets all public prayer requests by campus 1`] = `
Object {
  "data": Object {
    "campusPrayers": Array [
      Object {
        "campus": Object {
          "id": "Campus:be35f49307d7297989d3514be788ef2d",
          "name": "Anderson",
        },
        "id": "Prayer:559b23fd0aa90e81b1c023e72e230fa1",
        "text": "Pray this works.",
      },
      Object {
        "campus": null,
        "id": "Prayer:965b6e6d7046a885bea4e300b5c0400d",
        "text": "Help me",
      },
    ],
  },
}
`;

exports[`Prayer resolver gets all saved prayers 1`] = `
Object {
  "data": Object {
    "savedPrayers": Array [
      Object {
        "id": "Prayer:559b23fd0aa90e81b1c023e72e230fa1",
        "text": "Pray this works.",
      },
      Object {
        "id": "Prayer:965b6e6d7046a885bea4e300b5c0400d",
        "text": "Help me",
      },
    ],
  },
}
`;

exports[`Prayer resolver increments prayed for a request 1`] = `
Object {
  "data": Object {
    "incrementPrayerCount": Object {
      "id": "Prayer:559b23fd0aa90e81b1c023e72e230fa1",
      "text": "Pray this works.",
    },
  },
}
`;

exports[`Prayer resolver saves a prayer 1`] = `
Object {
  "data": Object {
    "savePrayer": Object {
      "id": "Prayer:559b23fd0aa90e81b1c023e72e230fa1",
      "text": "Pray this works.",
    },
  },
}
`;

exports[`Prayer resolver unsaves a prayer 1`] = `
Object {
  "data": Object {
    "unSavePrayer": Object {
      "id": "Prayer:559b23fd0aa90e81b1c023e72e230fa1",
      "text": "Pray this works.",
    },
  },
}
`;
